File: actions/fulfillment/sync.go
package fulfillment

// SyncInput of the fulfillmentRequest
type SyncInput struct {
	Intent string `json:"intent" binding:"required"`
}

// SyncPayload payload for Response
type SyncPayload struct {
	// Reflects the unique (and immutable) user ID on the agent's platform. The string is opaque to Google, so if there's an immutable form vs a mutable form on the agent side, use the immutable form (e.g. an account number rather than email).
	AgentUserID string `json:"agentUserId" binding:"required"`
	// For systematic errors on SYNC
	ErrorCode string `json:"errorCode"`
	// Detailed error which will never be presented to users but may be logged or used during development.
	DebugString string `json:"debugString"`
	// List of devices owned by the user. Zero or more devices are returned (zero devices meaning the user has no devices, or has disconnected them all).
	Devices []Device `json:"devices" binding:"required"`
}

// SyncResponse response for fulfillment request
type SyncResponse struct {
	// ID of the corresponding request.
	RequestID string `json:"requestId" binding:"required"`
	// Intent response payload.
	Payload SyncPayload `json:"payload" binding:"required"`
}

// SyncRequest is request type
type SyncRequest struct {
	RequestID string      `json:"requestId" binding:"required"`
	Inputs    []SyncInput `json:"inputs" binding:"required"`
}

File: fulfillment/src/intent/sync.rs
use serde::Serialize;
use uuid::Uuid;
use houseflow_db::{Database, models::{User, Device}};
use crate::Error;

// Empty module, SYNC intent doesn't have any payload
pub mod request {
}

pub mod response {
    use super::*;

    #[derive(Serialize)]
    pub struct Payload {
        /// Reflects the unique (and immutable) user ID on the agent's platform.
        #[serde(rename = "agentUserId")]
        pub user_id: Uuid,

        /// For systematic errors on SYNC
        #[serde(rename = "errorCode")]
        pub error_code: Option<String>,

        /// Detailed error which will never be presented to users but may be logged or used during development.
        #[serde(rename = "debugString")]
        pub debug_string: Option<String>,

        /// List of devices owned by the user.
        /// Zero or more devices are returned (zero devices meaning the user has no devices, or has disconnected them all).
        pub devices: Vec<Device>,
    }

    #[derive(Serialize)]
    pub struct Response {
        /// ID of the request.
        #[serde(rename = "requestId")]
        pub request_id: String,

        /// Intent response payload.
        pub payload: Payload,
    }
}

use response::Response;

pub async fn handle(db: &Database, user: &User, request_id: String) -> Result<Response, Error> {
    log::debug!("Received Sync intent from User ID: {}", user.id.to_string());
    let devices = db.get_user_devices(user.id).await?;
    
    Ok(Response {
        request_id,
        payload: response::Payload {
            user_id: user.id,
            error_code: None,
            debug_string: None,
            devices,
        }
    })
}
